include/ruby/internal/intern/process.h

```diff
@@ -28,7 +28,9 @@
 RBIMPL_SYMBOL_EXPORT_BEGIN()
 
 /* process.c */
-void rb_last_status_set(int status, rb_pid_t pid);
+RUBY_EXTERN void (* rb_socket_before_fork_func)();
+
+void rb_last_status_set(rb_pid_t pid, int status, int error);
```

Comment:
:-1: Please no.  Don't break existing codes.

Reply:
@nobu already fixed this, but things in `intern` are not public interface right?

Reply:
Everything placed under `include/` are public interface.

I know @matz didn't expect others to use things under `intern/` when he made the directory decades ago.  But people use them anyways.  We cannot think that directory being private today.

Reply:
The file path `include/ruby/internal/intern/process.h` sounds like internal, that's rather confusing.

Is the rule that whatever is between `RBIMPL_SYMBOL_EXPORT_BEGIN()` and `RBIMPL_SYMBOL_EXPORT_END()` public C API?

Reply:
Maybe #3079 was not a good idea...

Reply:
Ultimately, if `make install` installs a file under `/usr/local/include/`, then 3rd parties can look at its contents (== no way for us to stop them using it).  Before 3.0 everything written in such file were public APIs.

I admit I made things complicated in 3.0.  Now the rule is:
- Unless otherwise noted, everything that we install under `/usr/local/include` are public APIs.
- File names are not a part of our public API.
- Those symbols prefixed with `rbimpl_` or `RBIMPL_` are not the part of our public API.

So `RBIMPL_SYMBOL_EXPORT_BEGIN() ... RBIMPL_SYMBOL_EXPORT_END()` is not related.  For instance `RSTRING_PTR()` is now a function, and is placed outside of the export section, and is still one of our public API.

Reply:
Thanks, I think I get it now:
* Everything under `include/ruby` is make-installed and public API (except `rbimpl_*`/`RBIMPL_*`). But, filenames including `internal` in it must not be referenced by users, only `include/ruby/*.h` are allowed to be `#include`-d by C extensions.
  So the "internal" in `include/ruby/internal` means "internal file name, do not rely on it" (and not internal APIs).
* Internal headers (= with only internal APIs) which are not make-installed are simply under `$repo_root/internal/` or directly under the root `$repo_root/*.h`.